### ISSUE TITLE
Last OPL music note keeps playing forever on core reboot 

### ISSUE THREAD

**Issue by vanfanel** [MiSTer-devel/ao486_MiSTer#174]
Hello there,

I noticed that, as the tittle says, if a program is playing an OPL music note an the core is reboot, that last note keeps sounding forever, until a new program resets the OPL.

@humply Can you see this bug too?

**Comment by vanfanel** [comment]
@humply I see! What you are doing is awesome enough, the core has me totally hooked since I discovered your improvements yesterday.. I can't stop trying games and demos on it!

@gtaylormb Maybe you know what's going on with the hanging OPL note on reboot?

**Comment by humply** [comment]
Yes, it's a great core, I know the feeling you're describing! I also noticed some music issues with [Cute Demo](https://github.com/mills32/CUTE_DEMO-MS-DOS/releases), but I'm not sure if the issues are with the code or with the AO486 core. You might want to try it out (if you run it with an older version of the AO486 core you can see some of the VGA scrolling issues that should now be fixed).

**Comment by gtaylormb** [comment]
What version of the core are you using?

In reducing the core logic, all state is now stored in memory, which does not get reset like registers do. At one point I added a state machine around the envelope state memory that forces all envelopes into the DECAY state, so any notes playing on reset normally gracefully decay at their programmed decay rate DR. However, in rare cases, an operator may have been programmed with no DR. I'm slightly reluctant to fix this as it it adds an additional state machine and logic around the envelope value memory itself, and I've really tried to shrink the core down as much as possible. It's a pretty easy fix but pretty rare I believe. As soon as you enter the next game it would normally fix itself as most games will zero out the register map on startup and overwrite the values.

**Comment by vanfanel** [comment]
> What version of the core are you using? The stable build is older; you may want to try an unstable nightly build after July.
> 
> In reducing the core logic, all state is now stored in memory, which does not get reset like registers do. At one point I added a state machine around the envelope state memory that forces all envelopes into the RELEASE state, so any notes playing on reset normally gracefully decay at their programmed release rate RR. However, in rare cases, an operator may have been programmed with no RR. I'm slightly reluctant to fix this as it it adds an additional state machine and logic around the envelope value memory itself, and I've really tried to shrink the core down as much as possible. It's a pretty easy fix but pretty rare I believe. As soon as you enter the next game it would normally fix itself as most games will zero out the register map on startup and overwrite the values.

Thanks for the detailed answer, it was very clear and has interesting information on the issue.

I am using this build, which I did from the latest sources on the repository, which were recently updated with many awesome fixes by @humply:
https://github.com/user-attachments/files/18210063/ao486.rbf.zip

I understand the OPL notes should "decay" on reset, and I would say it happened at some previous version of the core, but it doesn't seem to happen anymore. Maybe related to those recent changes?

**Comment by gtaylormb** [comment]
> I understand the OPL notes should "decay" on reset, and I would say it happened at some previous version of the core, but it doesn't seem to happen anymore. Maybe related to those recent changes?

This is on my to-do list to debug and fix if I can replicate.

>I wonder if the clock domain crossing causes the last command to be replayed on core reboot?

I don't believe so.

**Comment by vanfanel** [comment]
@gtaylormb Were you able to reproduce the "last OPL note standing on reset" bug? :P

**Comment by gtaylormb** [comment]
> [@gtaylormb](https://github.com/gtaylormb) Were you able to reproduce the "last OPL note standing on reset" bug? :P

Yes, I've replicated the issue and confirmed it was a regression in e881af1523d958e0dcda4f06f99a95deedcf717b. I'll work on a fix

**Comment by vanfanel** [comment]
@gtaylormb This is not fixed, sorry.
This is the core build I'm using:
https://github.com/MiSTer-unstable-nightlies/ao486_MiSTer/releases/download/unstable-builds/ao486_unstable_20250224_205134.rbf

Use Prince of Persia, for example, let the tittle screen music play and give the core a CTRL+ALT+SUP: you will still hear part of the note plays forever after the system reboots.

**Comment by gtaylormb** [comment]
Probably that note is configured with a release rate of 0 I'm guessing. I'm pretty busy for the next few days but I'll replicate your testcase and work on a full fix. I also do not have permission to reopen.

**Comment by gtaylormb** [comment]
> Only reset via CTRL+ALT+SUP is affected: resetting via the menu the problem does not appear.

This is interesting. I wonder if resetting this way asserts the reset signal to the core at all, in which case, there's not much I can do. But I'll definitely test and find out.

**Comment by sorgelig** [comment]
if CTRL+ALT+SUP means CTRL+Alt+DEL then reset signal should be this one:
https://github.com/MiSTer-devel/ao486_MiSTer/blob/55fe09ce88b767a47a6f93d87b9ef26db9a36c26/rtl/system.v#L38

**Comment by starquake** [comment]
Just as an aside, and sorry if this is noise. But I remember from back in the day that sometimes on a real machine when a game crashed and caused a reboot the real OPL3 also could become stuck and would keep sounding a note :-P

**Comment by gtaylormb** [comment]
I also remember this with my Pro Audio Spectrum 16 :)

Will trace this signal through and see if it passes through to the reset signal of the OPL3 when I get a chance.

**Comment by gtaylormb** [comment]
Confirmed the notes keep playing on CTRL-ALT-DEL and that the reset is not connected internally to `ps2_reset_n`, so it is not reaching the OPL3 core it seems. The behavior is not the same as a hard reset (pressing RESET button).

Not sure what the intention of the reset configuration is. `ps2_reset_n` is connected to the `BUTTONS` output at the top level. There is also a `buttons` signal (different case, different signal), and this one appears to flow through to `reset`:

https://github.com/MiSTer-devel/ao486_MiSTer/blob/55fe09ce88b767a47a6f93d87b9ef26db9a36c26/ao486.sv#L858

Seems if we also | in `BUTTONS[1]` this would fix, but not sure of the other effects this has because this reset goes everywhere.

**Comment by sorgelig** [comment]
Connected ps2_reset_n  to BUTTONS[1] makes it acts as USER button, i.e. core reset. BUTTONS goes to Main and processed as hardware OSD and USER buttons. Main returns it in `buttons[]` signal.

**Comment by sorgelig** [comment]
MIDI get correct reset by CTRL+ALT+DEL, so i think it's something specific to OPL3 implementation.

**Comment by gtaylormb** [comment]
Confirmed `reset` does not hit the core on CTRL-ALT-DEL, but it does on hitting the reset button itself.

**Comment by vanfanel** [comment]
> Confirmed `reset` does not hit the core on CTRL-ALT-DEL, but it does on hitting the reset button itself.

So.. would It be possible to have RESET hit the core via CRTL-ALT-DEL, or am I misinderstanding something about the issue?

**Comment by gtaylormb** [comment]
> So.. would It be possible to have RESET hit the core via CRTL-ALT-DEL, or am I misunderstanding something about the issue?

There is probably a way to hook into a reset that is occurring at CTRL-ALT-DEL. I was at the point where a SignalTap into the reset signals at the top level to get some visibility would be helpful.

**Comment by vanfanel** [comment]
> > So.. would It be possible to have RESET hit the core via CRTL-ALT-DEL, or am I misunderstanding something about the issue?
> 
> There is probably a way to hook into a reset that is occurring at CTRL-ALT-DEL. I was at the point where a SignalTap into the reset signals at the top level to get some visibility would be helpful.

Maybe Sorgelig can tell you how to do it?

**Comment by sorgelig** [comment]
if you are talking about `ps2_reset_n` signal, then it's wired to USER button which is HW reset for core:
```
assign BUTTONS     = {~ps2_reset_n, 1'b0};
```

**Comment by vanfanel** [comment]
The PCXT core had this same bug, but it was fixed with this commit:
https://github.com/MiSTer-devel/PCXT_MiSTer/commit/864d32ca95f345a0f3450d27ecc24aa6c2cd4e5f

@humply Maybe you can take a look at this with the mentioned commit on the XT core, please?

**Comment by sorgelig** [comment]
on windows ctrl+alt+del doesn't execute reset. Also there is a programmatical reset not considered in that commit.

**Comment by vanfanel** [comment]
> on windows ctrl+alt+del doesn't execute reset. Also there is a programmatical reset not considered in that commit.

Thing is, this core is intended to run MS-DOS if I am not mistaken, not Windows (which it runs in some form or another I suppose, but it's not the main OS it's supposed to run) and MS-DOS does reset on CTRL+ALT+SUP unless a certain interrupt vector is taken over.

In other words, reset on CTRL+ALT+SUP is the expected behavior on MS-DOS, minus exceptions.

**Comment by humply** [comment]
The following MiSTer-level reset methods should (non-exhaustive list) send a reset signal to the ao486 core and thus also to the OPL:
- Pressing LCtrl + LAlt + RAlt on the keyboard to reset the core
- Selecting "Reset and apply HDD" from the MiSTer menu
- Pressing the physical "User" button on the MiSTer device

How to trigger the hanging note issue:
- Reset the core while OPL music is playing, for example:
  - During the Dune II intro music (Westwood logo)
  - During the Prince of Persia intro music

Each time the core is reset again using any of the methods listed above while a hanging note is playing, that note will replay from the beginning and then hang again. The fact that the note replays suggests that the OPL is receiving a reset signal.

However, pressing Ctrl + Alt + Del also triggers the hanging note issue, but when pressed while a hanging note is already playing, it has no effect on the hanging note. The note simply continues to hang (no replay). So Ctrl + Alt + Del does not currently trigger an OPL reset.

Contrary to @vanfanel's earlier claim that only a specific reset method was affected, I can reproduce the hanging note issue with any reset method.

As @gtaylormb explained earlier, the OPL's internal state is likely stored in memory, and that memory isn't reset like the registers are when the OPL receives a reset.

Therefore, the problem seems not to be with the reset signal itself (as @sorgelig also suggested), but rather that the OPL, upon receiving the reset signal, doesn't clear its internal memory, leaving hanging notes stuck.

It's unclear how physical sound cards handled the ISA reset signal, but the observed difference between a Ctrl + Alt + Del reset and other reset methods is probably expected behavior for a warm reboot. (Does Ctrl + Alt + Del trigger the `ps2_reset_n`?)
`~ps2_reset_n` -> `BUTTONS[1]` -> `Main_MiSTer` -> `hps_io` -> `buttons[1]`

https://github.com/MiSTer-devel/ao486_MiSTer/blob/42bf8829df77d65be3e1a2a1dd2b22a2a2aa0123/ao486.sv#L864

There are software-based solutions that can reset the OPL and resolve the hanging note issue. This supports the idea that something in the OPL isn't being properly cleared on reset.
- https://github.com/stargo/resetopl/releases

@gtaylormb Commenting out specific lines in the `resetopl` source code might help identify which part of the OPL needs to be reset to resolve hanging notes.

The hanging note issue has also been observed on real hardware, so it's hardware-accurate, but I understand the "quality of life" improvement argument. 😄

After reviewing the PCXT core commit referenced earlier by @vanfanel, I agree with @sorgelig that adding hardware-level detection of the Ctrl + Alt + Del key combination would not be hardware-accurate. The given Windows example illustrates that tying a particular key combination to a hardware component reset isn't correct behavior and would introduce issues not seen on real hardware. The existing MiSTer-level reset methods should be sufficient for resetting the OPL, ideally also clearing any hanging notes.

Additionally, the PCXT OPL implementation differs from the ao486 one. In the ao486 core, the OPL does receive the reset signal on, for example, LCtrl + LAlt + RAlt, so the referenced PCXT commit wouldn't resolve the ao486 hanging note issue.

**Conclusion**
The OPL seems to receive a reset signal when a MiSTer-level core reset occurs, but the OPL reset does not currently clear hanging notes.

**Comment by sorgelig** [comment]
May be, after receiving the reset, OPL code can run a loop writing default values (probably 0 will be ok) to each OPL register?

**Comment by gtaylormb** [comment]
>The hanging note issue has also been observed on real hardware, so it's hardware-accurate, but I understand the "quality of life" improvement argument. 😄

I've been a bit reluctant to dig into this because this used to happen on my own system, so it's a bit nostalgic for me. We're adding extra logic to support a feature that is not accurate. It's tempting to close this request with will not fix, but the consensus seems to be that we should "fix."

I'll try to spend some time over the next few days working on this. There's some other mods I want to get into the OPL3 anyway--one potential bug (have not verified yet), and some code style changes and lint fixes I think would improve the readability.

@sorgelig how do you feel about git submodules? Really tempting to use here.

**Comment by humply** [comment]
Thank you for your response and for the great improvements you've made to this core! I just wanted to share my perspective on this "issue", and clarify a few things about the possible cause. I didn't intend to suggest that you should fix or work on this. Your current OPL implementation is already excellent and a huge improvement over the previous one. Personally, I don't mind the hanging note behavior, as I remember it occurring on real hardware as well. :-) However, other users might see it as a bug, so if it ever proves fixable, perhaps a "Hanging note" toggle in the menu could be an option? I mainly wanted to share my thoughts on where the behavior might originate (though I could certainly be wrong). I'm also not sure how feasible it is to have the OPL reset clear any hanging notes or if the current OPL reset behavior should even be altered.

As for the hardware side, I believe some sound cards eventually addressed this hanging note issue, though I'm not entirely sure. I can try to test it on an AWE64 I should still have somewhere.

I ran into a similar dilemma with the PIT and with the ET4000 VGA lately and ended up tweaking some logic to improve compatibility. I'm still relatively new to FPGA development, and it's been fascinating to be able to make hardware level changes. It's also interesting to realize how many quirks old hardware actually had, and how to balance accuracy against compatibility. PC hardware was very diverse, so it's difficult to set requirements or goals for this core.

**Comment by sorgelig** [comment]
@gtaylormb Github made submodules extremely inconvenient. Also submodule doesn't solve any problems but adds more obfuscation where it takes more time to see the changes. Sometimes external repositories disappear or moved making maintenance even harder. Sometimes specific change need to be made for different reasons which is not good for original module or author is not reachable or refuse to add the changes, etc.. So i'm against submodules. In some special cases submodule could be used but repository of submodule must be in MiSTer-devel to make sure the integrity.
It's not a problem to add a credit and link to your repo. Probably it's even better to add it in a header of the source file.

**Comment by gtaylormb** [comment]
@humply thanks for the feedback.

@sorgelig makes sense. Yeah it's not about credit it's just about me having to manually port changes between two repos, but I understand your reasoning. When you have no control over the external repo I can see this being a huge pain.

**Comment by sorgelig** [comment]
Resetting from some games (like Cyber Sphere) guaranteed to get a noise/whistle, so adding HW reset to OPL inside the core could be a very welcome. I don't think anyone would miss this issue presented in real HW :)

Before that, mentioned earlier `resetopl.exe` can be added into autoexec as a workaround.
